src/NHibernate/Impl/ExpressionProcessor.cs

```diff
@@ -248,32 +248,73 @@ private static ICriterion Le(ProjectionInfo property, object value)
 		}
 
 		/// <summary>
-		/// Invoke the expression to extract its runtime value
+		/// Walk or Invoke expression to extract its runtime value
 		/// </summary>
 		public static object FindValue(Expression expression)
 		{
-			if (expression.NodeType == ExpressionType.Constant)
-				return ((ConstantExpression) expression).Value;
-
-			if (expression.NodeType == ExpressionType.MemberAccess)
+			object value;
+			switch (expression.NodeType)
 			{
-				var memberAccess = (MemberExpression) expression;
-				if (memberAccess.Expression == null || memberAccess.Expression.NodeType == ExpressionType.Constant)
-				{
-					var constantValue = ((ConstantExpression) memberAccess.Expression)?.Value;
-					var member = memberAccess.Member;
-					switch (member.MemberType)
+				case ExpressionType.Constant:
+					var constantExpression = (ConstantExpression) expression;
+					return constantExpression.Value;
+				case ExpressionType.MemberAccess:
+					var memberExpression = (MemberExpression) expression;
+					value = memberExpression.Expression != null ? FindValue(memberExpression.Expression) : null;
+
+					switch (memberExpression.Member.MemberType)
 					{
 						case MemberTypes.Field:
-							return ((FieldInfo) member).GetValue(constantValue);
+							return ((FieldInfo) memberExpression.Member).GetValue(value);
 						case MemberTypes.Property:
-							return ((PropertyInfo) member).GetValue(constantValue);
+							return ((PropertyInfo) memberExpression.Member).GetValue(value);
 					}
-				}
+					break;
+				case ExpressionType.Call:
+					var methodCallExpression = (MethodCallExpression) expression;
+					var args = new object[methodCallExpression.Arguments.Count];
+					for (int i = 0; i < args.Length; i++)
+						args[i] = FindValue(methodCallExpression.Arguments[i]);
+
+					if (methodCallExpression.Object == null) //extension or static method
+					{
+						if (args.Length > 0 && args[0] != null)
+						{
+							return methodCallExpression.Method.Invoke(args[0].GetType(), args);
+						}
+						else
+						{
+							return methodCallExpression.Method.Invoke(methodCallExpression.Method.DeclaringType, args);
+						}
+					}
+					else
+					{
+						var callingObject = FindValue(methodCallExpression.Object);
+
+						return methodCallExpression.Method.Invoke(callingObject, args);
+					}
+				case ExpressionType.Convert:
+					var unaryExpression = (UnaryExpression) expression;
+					value = FindValue(unaryExpression.Operand);
+					var type = Nullable.GetUnderlyingType(unaryExpression.Type) ?? unaryExpression.Type;
+					if (type == typeof(object) || value == null)
```

Comment:
`|| value == null`

So it seems new code will happily process the following query:
```C#
object param = null;
query.Where(x => x.Id == (int)param);
```

Make sure it behaves the same with and without optimization.

Reply:
Convert is to much trick, Convert.ChangeType fails with invalid convert tests so I try to avoid cast as much as possible.

Reply:
Convert might need some additional looking into for more optimizations. But seems safe enough for a start...